src/main/jbake/content/blog/2017/06/23/opentracing-appmetrics.adoc

```diff
@@ -0,0 +1,270 @@
+deployed within https://kubernetes.io/[Kubernetes].
+
+== The example application
+
+In a recent article we showed how a Spring Boot application could easily be
+http://www.hawkular.org/blog/2017/06/9/opentracing-spring-boot.html[instrumented using OpenTracing].
+
+The https://github.com/objectiser/opentracing-prometheus-example[example] we are going to use in
+this article uses the same approach to create two services, _service1_ and _service2_ (sorry, not very
+imaginative names).
+
+_Service2_ presents a single REST endpoint (`/hello`) for internal use by _Service1_. The code for this endpoint
+is:
+
+.Service 2's Hello Controller:
+----
+    @Autowired
+    private io.opentracing.Tracer tracer;
+
+    @RequestMapping("/hello")
+    public String hello() throws InterruptedException {
+        Thread.sleep(1 + (long)(Math.random()*500));
+        if (Math.random() > 0.8) {
+            tracer.activeSpan().setTag("error", true);
+        }
+        return "Hello from Spring Boot!";
+    }
+----
+
+The first line simply introduces a random delay, to make the collected metrics more interesting. The next three
+lines randomly cause the span (associated with the REST endpoint invocation) being tagged as an error.
+
+_Service1_ presents three REST endpoints for use by an end user. These are:
+
+.Service 1's Txn Controller:
+----
+    @RequestMapping("/buy")
+    public String buy() throws InterruptedException {
+        Thread.sleep(1 + (long)(Math.random()*500));
+        tracer.activeSpan().setBaggageItem("transaction", "buy");
+        ResponseEntity<String> response = restTemplate.getForEntity(service2Url + "/hello", String.class);
+        return "BUY + " + response.getBody();
+    }
+
+    @RequestMapping("/sell")
+    public String sell() throws InterruptedException {
+        Thread.sleep(1 + (long)(Math.random()*500));
+        tracer.activeSpan().setBaggageItem("transaction", "sell");
+        ResponseEntity<String> response = restTemplate.getForEntity(service2Url + "/hello", String.class);
+        return "SELL + " + response.getBody();
+    }
+
+    @RequestMapping("/fail")
+    public String fail() throws InterruptedException {
+        Thread.sleep(1 + (long)(Math.random()*500));
+        ResponseEntity<String> response = restTemplate.getForEntity(service2Url + "/missing", String.class);
+        return "FAIL + " + response.getBody();
+    }
+----
+
+All three methods introduce a random delay.
+
+The `buy` and `sell` methods additional set a baggage item `transaction` with the name of the business
+transaction being performed (i.e. buy or sell). For those not familiar with OpenTracing, the
+https://github.com/opentracing/specification/blob/master/specification.md#set-a-baggage-item[baggage concept]
+allows information to be carried _in band_ with the trace context between invoked services. Shortly we will show
+how this can be used to isolate the metrics relevant only for a particular business transaction.
+
```

Comment:
“Shortly” sounds odd to me, as long as we’re going to show how to do it in this article, I think we can delete that word.

"We will show you how a baggage item can be used to isolate the metrics relevant only for a particular business transaction."